basalt/autograd/ops/basics.mojo

```diff
@@ -711,3 +711,52 @@ struct RESHAPE:
         memcpy(res_grad.data(), ug.data(), ug_shape.num_elements())
 
         return res_grad ^
+
+struct FMA:
+    @staticmethod
+    fn result_shape(t1_shape: TensorShape, t2_shape: TensorShape, t3_shape: TensorShape) -> TensorShape:
+        return broadcast_shapes(t1_shape, t2_shape, t3_shape)
+
+    @staticmethod
+    fn forward[
+        t1_shape: TensorShape,
+        t2_shape: TensorShape,
+        t3_shape: TensorShape,
+    ](inout res: Tensor[dtype], t1: Tensor[dtype], t2: Tensor[dtype], t3: Tensor[dtype]):
+        """
+        Forward pass of the fma operation.
+        """
+        
+        @parameter
+        fn vec_fma[Nelts: Int](i: Int):
+            res.store[Nelts](
+                i,
+                t1.load[Nelts](i).fma(t2.load[Nelts](i), t3.load[Nelts](i))
+            )
+
+        vectorize[vec_fma, nelts, size=t1_shape.num_elements()]()
+
+    @staticmethod
+    fn backward[
+        tensor_id: Int,
+        ug_shape: TensorShape,
+        t1_shape: TensorShape,
+        t2_shape: TensorShape,
+        t3_shape: TensorShape,
+    ](ug: Tensor[dtype], t1: Tensor[dtype], t2: Tensor[dtype], t3: Tensor[dtype]) -> Tensor[dtype]:
+        """Backward operation of fma."""
+        # d(x * y + z) / dx = y
+        # d(x * y + z) / dy = x
+        # d(x * y + z) / dz = 1
+
+        @parameter
+        if tensor_id == 0:
+            var res_grad = Tensor[dtype](ug_shape)
+            elwise_op[ug_shape, t2_shape, mul](res_grad, ug, t2)
+            return res_grad ^
+        elif tensor_id == 1:
+            var res_grad = Tensor[dtype](ug_shape)
+            elwise_op[ug_shape, t1_shape, mul](res_grad, ug, t1)
+            return res_grad ^
+        else:
```

Comment:
Should we put a `constrained[False, ...]()` here?

Reply:
Maybe, but I do think we need a wider discussion about how we want to manage Error handling, especially when we get to more constrained operators like Cast

Reply:
For `cast`, I thin constrain can be imposed like `stdlib.SIMD`. `constrained[False]()` is really poor man's unreachable. I'm good with not addressing error handling here.